rgw: move forward_request_to_master() out of sal

cbodley opened the pull request:
bucket creation/deletion were the only things that used it inside of sal. it's now the caller's responsibility to forward these requests before calling into sal. in order for `RGWBucketCreate` to call `forward_request_to_master()` itself, other things like bucket placement needed to move out as well

renamed the `sal::Driver::get_bucket()` functions that load the bucket from storage to `load_bucket()`. the non-loading `get_bucket()` function never fails, so it now returns `std::unique_ptr<Bucket>` directly

because `RGWBucketCreate` always has to handle the existing-bucket case, i moved the `create_bucket()` function from `sal::User` into `sal::Bucket::create()`. that allows us to reuse the handle we allocated in `sal::Driver::load_bucket()`

finally, moves `Driver::forward_request_to_master()` to a free function in rgw_op.cc, and `Driver::forward_iam_request_to_master()` to a free function in rgw_rest_role.cc

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

cbodley commented:
fixed up the dbstore tests now that the caller no longer passes in an objv to `create_bucket()`

cbodley commented:
passed qa after rerun https://pulpito.ceph.com/cbodley-2023-05-15_13:00:49-rgw-wip-cbodley-testing-distro-default-smithi/

yuriw commented:
@cbodley needs rebase and merge

dang commented on src/rgw/rgw_sal.h:
```diff
@@ -673,6 +652,27 @@ class Bucket {
+    };
+
+    /// Create this bucket in the backing store.
+    virtual int create(const DoutPrefixProvider* dpp,
```
Minor nit: Bucket now has create() and remove_bucket().  I think, for consistency, this should be create() and remove() or create_bucket() and remove_bucket().

cbodley replied:
any preference? if not, i'll go with create() and remove()

dang commented on src/rgw/rgw_sal.h:
```diff
@@ -289,21 +290,16 @@ class Driver {
-    /** Lookup a Bucket by name.  Queries driver for bucket info. */
-    virtual int get_bucket(const DoutPrefixProvider* dpp, User* u, const std::string& tenant, const std::string& name, std::unique_ptr<Bucket>* bucket, optional_yield y) = 0;
+    virtual std::unique_ptr<Bucket> get_bucket(User* u, const RGWBucketInfo& i) = 0;
+    /** Load a Bucket by key.  Queries driver for bucket info. */
```
I think the comment should say that *bucket is valid on -EEXIST, so implementers know (and I don't forget)

cbodley commented:
addressed comments, and also removed the load_bucket() overload that took `string tenant, string bucket_name` and converted everything to use the `rgw_bucket` overload

@dang i did some light testing of the posix changes with s3tests, but it's hard for me to tell whether or not the failures were regressions. do you want to do any extra testing of this posix stuff?

dang commented on src/rgw/rgw_lc.cc:
```diff
@@ -1525,7 +1521,9 @@ int RGWLC::bucket_lc_process(string& shard_id, LCWorker* worker,
 
-  int ret = driver->get_bucket(this, nullptr, bucket_tenant, bucket_name, &bucket, null_yield);
+  int ret = driver->load_bucket(this, nullptr,
+                                rgw_bucket_key(bucket_tenant, bucket_name),
```
should this be rgw_bucket rather than rgw_bucket_key?

cbodley replied:
good catch, thanks

dang reviewed (approved):
I'll do final testing when I rebase onto this.  The changes are minor, and shouldn't introduce any regressions.

cbodley commented:
[latest qa run](https://pulpito.ceph.com/cbodley-2023-10-06_14:37:21-rgw-wip-rgw-sal-forward-distro-default-smithi/) had a `RGWLC::process()` crash against dbstore, and a tempest failure on `tempest.api.object_storage.test_account_bulk.BulkTest.test_extract_archive`

cbodley commented:
> [latest qa run](https://pulpito.ceph.com/cbodley-2023-10-06_14:37:21-rgw-wip-rgw-sal-forward-distro-default-smithi/) had a `RGWLC::process()` crash against dbstore

i see the same dbstore+lifecycle crashes on main, so i'm not sure why it only shows up on this branch

> and a tempest failure on `tempest.api.object_storage.test_account_bulk.BulkTest.test_extract_archive`

`RGWBulkUploadOp::handle_dir()` was missing a call to `select_bucket_placement()`

cbodley commented:
now `tempest.api.object_storage.test_account_bulk.BulkTest.test_extract_archive` is [failing](http://qa-proxy.ceph.com/teuthology/cbodley-2023-11-03_13:48:00-rgw-wip-rgw-sal-forward-distro-default-smithi/7445529/teuthology.log) to list an object from the archive

cbodley commented:
> now `tempest.api.object_storage.test_account_bulk.BulkTest.test_extract_archive` is [failing](http://qa-proxy.ceph.com/teuthology/cbodley-2023-11-03_13:48:00-rgw-wip-rgw-sal-forward-distro-default-smithi/7445529/teuthology.log) to list an object from the archive

the [rgw log](http://qa-proxy.ceph.com/teuthology/cbodley-2023-11-03_13:48:00-rgw-wip-rgw-sal-forward-distro-default-smithi/7445529/remote/smithi002/log/rgw.ceph.client.0.log.gz) seems to think the bucket created by bulk-upload only has a single shard, so doesn't find the object inside:
> swift:list_bucket cls_bucket_list_ordered: request from each of 1 shard(s) for 10001 entries to get 10001 total entries

when i tried to reproduce this under vstart, it had the expected 11 shards:
> swift:list_bucket cls_bucket_list_ordered: request from each of 11 shard(s) for 976 entries to get 10001 total entries

the reproducer, including the magic curl command (where X-Auth-Token was copied from `swift list --debug` output):
```
# contents of archive.tar:
$ tar -tf archive.tar
archive/
archive/foo.txt

# upload archive.tar to the account root:
$ curl -i http://localhost:8000/swift/v1?extract-archive=tar -X PUT -H "X-Auth-Token: AUTH_rgwtk0b000000746573743a746573746572b7c23cb32a2810d1d7824a654d45c31f67ffa8be2fd012232586f398831746431d7dfe8a" -H "Content-Length: 10240" --data-binary "@archive.tar"
HTTP/1.1 200 OK
Transfer-Encoding: chunked
X-Trans-Id: tx000001049b8c5eedd0d97-0065494b4b-1033-default
X-Openstack-Request-Id: tx000001049b8c5eedd0d97-0065494b4b-1033-default
Content-Type: text/plain; charset=utf-8
Server: Ceph Object Gateway (reef)
Date: Mon, 06 Nov 2023 20:23:45 GMT
Connection: Keep-Alive

Number Files Created: 1
Response Body: 
Response Status: 201 Created
Errors: 

# list the extracted 'archive' container:
$ swift -A http://localhost:8000/auth/v1 -U test:tester -K testing list archive
foo.txt
```

pritha-srivastava commented on src/rgw/rgw_rest_role.cc:
```diff
@@ -13,15 +13,68 @@
 
 using namespace std;
 
+int forward_iam_request_to_master(const DoutPrefixProvider* dpp,
```
@cbodley : Just a question here - if in future we plan to add multisite replication for `oidc-provider` in rgw, where should this method be shifted to? `rgw_rest_iam.h`?

cbodley replied:
rgw_rest_conn.* might be a good place since these functions depend on RGWRESTConn

cbodley commented:
> > now `tempest.api.object_storage.test_account_bulk.BulkTest.test_extract_archive` is [failing](http://qa-proxy.ceph.com/teuthology/cbodley-2023-11-03_13:48:00-rgw-wip-rgw-sal-forward-distro-default-smithi/7445529/teuthology.log) to list an object from the archive

ok, [fixed](http://qa-proxy.ceph.com/teuthology/cbodley-2023-11-07_19:06:23-rgw-wip-rgw-sal-forward-distro-default-smithi/7450822/teuthology.log) by adding the `createparams.zone_placement = rgw::find_zone_placement(...)` to `RGWBulkUploadOp::handle_dir()`. will rerun the whole suite now

cbodley commented:
jenkins test api

cbodley commented:
qa in https://pulpito.ceph.com/cbodley-2023-11-07_19:46:35-rgw-wip-rgw-sal-forward-distro-default-smithi/ hit a strange `test_s3select.py::test_column_sum_min_max` failure in rgw/crypt/barbican that i haven't seen before, but it went away with a rerun https://pulpito.ceph.com/cbodley-2023-11-08_13:38:09-rgw-main-distro-default-smithi/, so i'm calling this a pass

cbodley commented:
oops, that rerun didn't use the right ceph branch. re-rerunning in https://pulpito.ceph.com/cbodley-2023-11-08_17:26:47-rgw-wip-rgw-sal-forward-distro-default-smithi/

cbodley commented on src/rgw/rgw_rest_role.cc:
```diff
@@ -13,15 +13,68 @@
+  }
+
+  // use the master zone's endpoints
+  auto conn = RGWRESTConn{dpp->get_cct(), z->second.id, z->second.endpoints,
+                          std::move(creds), zg->second.id, zg->second.api_name};
+  bufferlist outdata;
+  constexpr size_t max_response_size = 128 * 1024; // we expect a very small response
+  int ret = conn.forward_iam_request(dpp, creds, req, nullptr, max_response_size,
+                                     &indata, &outdata, y);
```
oops, `creds` was moved into the `RGWRESTConn` ctor call above. i opened https://github.com/ceph/ceph/pull/55148 to address this

cc @jzhu116-bloomberg
